src/util.c

```diff
@@ -552,6 +552,30 @@ int string2d(const char *s, size_t slen, double *dp) {
     return 1;
 }
 
+/* Returns 1 if the double value can safely be represented in long long without
+ * precision loss, in which case the corresponding long long is stored in the out variable. */
+int double2ll(double d, long long *out) {
+#if (DBL_MANT_DIG >= 52) && (LLONG_MAX == 0x7fffffffffffffffLL)
+        /* Check if the float is in a safe range to be casted into a
+         * long long. We are assuming that long long is 64 bit here.
+         * Also we are assuming that there are no implementations around where
+         * double has precision < 52 bit.
+         *
+         * Under this assumptions we test if a double is inside an interval
+         * where casting to long long is safe. Then using two castings we
+         * make sure the decimal part is zero. If all this is true we can use
+         * integer without precision loss. */
+        double min = -4503599627370495; /* -(2^52)+1 */
+        double max = 4503599627370495; /* (2^52)-1 */
+        if (d < min || d > max)
+            return 0;
```

Comment:
This is basically an optimization - it assumes the comparison of `d` to constant doubles is faster than the casting and comparison done below. We should benchmark this to make sure we aren't over complicating the code.

Reply:
I've just done a simple test and the performance doesn't seem to be any different.

Reply:
@sundb you mean you removed the `if` and let it always flow to the conversion?
such a test should be done on scores that are outside the range (i.e check if an early abort is faster than a late abort).
is that what you did?

Reply:
> @sundb you mean you removed the `if` and let it always flow to the conversion? 
    This is done, and I use the value out of the range (14503599627370495)

Reply:
Convert -14503599627370494(out of range) (with range check 1.1 faster than without range check)
https://quick-bench.com/q/v72hz6-wEKYnxE3BkogGhNK_PUE

Convert -4503599627370494(in range) (with range check 1.1 slower than without range check)
https://quick-bench.com/q/urJtjViZ2RtBzEIvW4oKdgGU3gg

Convert 4503599627370494(in range)  (with range check 1.1 slower than without range check)
https://quick-bench.com/q/a2vj2Gxlrp4iU-jTnhV2XoQwkuY

Reply:
looks like these benchmarks are with `-O0`, so not sure they worth much.

Reply:
Yes, I used O0 on purpose, I used O2 again for the same test, and they performed exactly the same(include https://github.com/redis/redis/pull/10486/commits/adc115043cedaa48355382ed8d3e3ca4ac52cbd9).
@zuiderkwast Your code is working, bringing a little boost.

Reply:
Nice. With O2, I think the benchmark isn't testing anything because the constant expression can be computed in compile time.

Reply:
@zuiderkwast Thanks for pointing out, I disabled the optimization of long long variables and tested it again.
Case 1:
Path: d < min
https://quick-bench.com/q/Xbsr1a_OJRyXEiMQBYGhM6do25o
Range check 1.1 faster than Conversion

Case 2:
Path: d > max
https://quick-bench.com/q/4TzA7nsFAb2qZuCTQqks73ldmBY
Range check 1.04 faster than Conversion

Reply:
i think it's pointless to benchmark things without optimizations, the compiler could generate completely different code.
the proper way to benchmark these is to trick the compiler not to completely optimize the whole thing out, i.e. use some conditional printf to print the output, but make sure it doesn't actually ever run, or something alike.

the other alternative is to just benchmark that inside redis, using a long pipeline, to see if it actually makes any measurable difference.

Reply:
My guess is that we'll not see any difference in a real Redis benchmark. Network and disk is much slower than ~3 clock cycles for this comparison vs cast. I suggest we keep the range check and not waste more time on this.